core/thread_flags.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2016 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     core_thread
+ * @{
+ *
+ * @file
+ * @brief       thread flags implementation
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+
+#include "bitarithm.h"
+#include "thread_flags.h"
+#include "irq.h"
+#include "thread.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#ifdef MODULE_CORE_THREAD_FLAGS
+static thread_flags_t _thread_flags_clear_atomic(thread_t *thread, thread_flags_t mask)
+{
+    unsigned state = irq_disable();
+    mask &= thread->flags;
+    thread->flags &= ~mask;
+    irq_restore(state);
+    return mask;
+}
+
+static void _thread_flags_wait(thread_flags_t mask, thread_t *thread, unsigned threadstate, unsigned irqstate)
+{
+    DEBUG("_thread_flags_wait: me->flags=0x%08x me->mask=0x%08x. going blocked.\n",
+            (unsigned)thread->flags, (unsigned)mask);
+
+    thread->wait_data = (void *)(unsigned)mask;
+    sched_set_status(thread, threadstate);
+    irq_restore(irqstate);
+    thread_yield_higher();
+}
+
+thread_flags_t thread_flags_clear(thread_flags_t mask)
+{
+    thread_t *me = (thread_t*) sched_active_thread;
+    mask = _thread_flags_clear_atomic(me, mask);
+    DEBUG("thread_flags_clear(): pid %"PRIkernel_pid" clearing 0x%08x\n", thread_getpid(), mask);
+    return mask;
+}
+
+static void _thread_flags_wait_any(thread_flags_t mask)
+{
+    thread_t *me = (thread_t*) sched_active_thread;
+    unsigned state = irq_disable();
+    if (!(me->flags & mask)) {
+        _thread_flags_wait(mask, me, STATUS_FLAG_BLOCKED_ANY, state);
+    }
+    else {
+        irq_restore(state);
+    }
+}
+
+thread_flags_t thread_flags_wait_any(thread_flags_t mask)
+{
+    thread_t *me = (thread_t*) sched_active_thread;
+    _thread_flags_wait_any(mask);
+    return _thread_flags_clear_atomic(me, mask);
+}
+
+thread_flags_t thread_flags_wait_one(thread_flags_t mask)
+{
+    _thread_flags_wait_any(mask);
+    thread_t *me = (thread_t*) sched_active_thread;
+    unsigned tmp = me->flags & mask;
+    return _thread_flags_clear_atomic(me, thread_flags_clear(1 << bitarithm_lsb(tmp)));
```

Comment:
If you are trying to unset all bits, but not the least significant one, then you may use `tmp &= -tmp`, or `tmp &= ~tmp + 1` (equivalent) instead of `1 << bitarithm_lsb(tmp)`.